frontend/src/lib/k8s/api/v2/webSocket.ts

```diff
@@ -1,11 +1,398 @@
 import { getToken } from '../../../auth';
 import { getCluster } from '../../../cluster';
 import { BASE_HTTP_URL } from './fetch';
 import { makeUrl } from './makeUrl';
 
-const BASE_WS_URL = BASE_HTTP_URL.replace('http', 'ws');
+// Constants for WebSocket connection
+export const BASE_WS_URL = BASE_HTTP_URL.replace('http', 'ws');
+
+/**
+ * Multiplexer endpoint for WebSocket connections
+ * This endpoint allows multiple subscriptions over a single connection
+ */
+const MULTIPLEXER_ENDPOINT = 'wsMultiplexer';
+
+/**
+ * Message types for WebSocket communication between client and server
+ */
+interface WebSocketMessage {
+  /** Cluster identifier */
+  clusterId: string;
+  /** API resource path */
+  path: string;
+  /** Query parameters */
+  query: string;
+  /** User identifier for authentication */
+  userId: string;
+  /** Message type for subscription management */
+  type: 'REQUEST' | 'CLOSE' | 'COMPLETE';
+}
+
+/**
+ * WebSocket manager to handle connections across the application.
+ * Provides a singleton-like interface for managing WebSocket connections,
+ * subscriptions, and message handling. Implements connection multiplexing
+ * to optimize network usage.
+ */
+export const WebSocketManager = {
+  /** Current WebSocket connection instance */
+  socketMultiplexer: null as WebSocket | null,
+
+  /** Flag to track if a connection attempt is in progress */
+  connecting: false,
+
+  /** Flag to track if we're reconnecting after a disconnect */
+  isReconnecting: false,
+
+  /** Map of message handlers for each subscription path */
+  listeners: new Map<string, Set<(data: any) => void>>(),
+
+  /** Set of paths that have received a COMPLETE message */
+  completedPaths: new Set<string>(),
+
+  /** Map of active WebSocket subscriptions with their details */
+  activeSubscriptions: new Map<string, { clusterId: string; path: string; query: string }>(),
+
+  /** Map to track pending unsubscribe operations for debouncing */
+  pendingUnsubscribes: new Map<string, NodeJS.Timeout>(),
+
+  /**
+   * Creates a unique key for identifying WebSocket subscriptions
+   * @param clusterId - Cluster identifier
+   * @param path - API resource path
+   * @param query - Query parameters
+   * @returns Unique subscription key
+   */
+  createKey(clusterId: string, path: string, query: string): string {
+    return `${clusterId}:${path}:${query}`;
+  },
+
+  /**
+   * Establishes or returns existing WebSocket connection
+   * Handles connection lifecycle, reconnection, and subscription restoration
+   * @returns Promise resolving to WebSocket connection
+   */
+  async connect(): Promise<WebSocket> {
+    // Return existing connection if available
+    if (this.socketMultiplexer?.readyState === WebSocket.OPEN) {
+      return this.socketMultiplexer;
+    }
+
+    // Wait for existing connection attempt if in progress
+    if (this.connecting) {
+      return new Promise(resolve => {
+        const checkConnection = setInterval(() => {
```

Comment:
Yes, there are some issues with this polling approach. But making this change now won't be easy so added detailed comments regarding tradeoffs

Reply:
I wonder if this has a measurable impact on app performance? It would be good to compare a before and after using the performance tab in developer tools. This is my main concern.

Adding an max 100ms isn't as big a deal IMHO.